wdl-ast/src/experimental.rs

```diff
@@ -0,0 +1,392 @@
+            Some(Self(syntax))
+        } else {
+            None
+        }
+    }
+
+    fn syntax(&self) -> &SyntaxNode {
+        &self.0
+    }
+}
+
+/// Represents a comment token in the AST.
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct Comment(SyntaxToken);
+
+impl AstToken for Comment {
+    fn can_cast(kind: SyntaxKind) -> bool
+    where
+        Self: Sized,
+    {
+        kind == SyntaxKind::Comment
+    }
+
+    fn cast(syntax: SyntaxToken) -> Option<Self>
+    where
+        Self: Sized,
+    {
+        match syntax.kind() {
+            SyntaxKind::Comment => Some(Self(syntax)),
+            _ => None,
+        }
+    }
+
+    fn syntax(&self) -> &SyntaxToken {
+        &self.0
+    }
+}
+
+/// Represents a version statement in a WDL AST.
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct VersionStatement(SyntaxNode);
+
+impl VersionStatement {
+    /// Gets the version of the version statement.
+    pub fn version(&self) -> Version {
+        token(&self.0).expect("version statement must have a version token")
+    }
+}
+
+impl AstNode for VersionStatement {
+    type Language = WorkflowDescriptionLanguage;
+
+    fn can_cast(kind: SyntaxKind) -> bool
+    where
+        Self: Sized,
+    {
+        kind == SyntaxKind::VersionStatementNode
+    }
+
+    fn cast(syntax: SyntaxNode) -> Option<Self>
+    where
+        Self: Sized,
+    {
+        match syntax.kind() {
+            SyntaxKind::VersionStatementNode => Some(Self(syntax)),
+            _ => None,
+        }
+    }
+
+    fn syntax(&self) -> &SyntaxNode {
+        &self.0
+    }
+}
+
+/// Represents a version in the AST.
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct Version(SyntaxToken);
+
+impl AstToken for Version {
+    fn can_cast(kind: SyntaxKind) -> bool
+    where
+        Self: Sized,
+    {
+        kind == SyntaxKind::Version
+    }
+
+    fn cast(syntax: SyntaxToken) -> Option<Self>
+    where
+        Self: Sized,
+    {
+        match syntax.kind() {
+            SyntaxKind::Version => Some(Self(syntax)),
+            _ => None,
+        }
+    }
+
+    fn syntax(&self) -> &SyntaxToken {
+        &self.0
+    }
+}
+
+/// Represents an identifier token.
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct Ident(SyntaxToken);
+
+impl AstToken for Ident {
+    fn can_cast(kind: SyntaxKind) -> bool
+    where
+        Self: Sized,
+    {
+        kind == SyntaxKind::Ident
+    }
+
+    fn cast(syntax: SyntaxToken) -> Option<Self>
+    where
+        Self: Sized,
+    {
+        match syntax.kind() {
+            SyntaxKind::Ident => Some(Self(syntax)),
+            _ => None,
+        }
+    }
+
+    fn syntax(&self) -> &SyntaxToken {
+        &self.0
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use pretty_assertions::assert_eq;
+
+    use super::*;
+
+    #[test]
+    fn document_has_preamble_comments() {
+        let parse = Document::parse(
+            r#"
+## This is a preamble comment.
```

Comment:
@adthrasher @peterhuene another instance where we want to be careful about pedantic definitions (because we're making up the terms):

**(a) Are all comments within the preamble "preamble comments"? OR (b) are _only_ double-pound-sign comments "preamble comments"?**

(a) is more intuitive. But then we need a new term for double-pound-sign comments. "Doc comments"? Then the only "valid" doc comments would be preamble doc comments. I think double-pound-sign comments outside the preamble would still be referred to as doc comments, they just aren't supported in any special way (and produce warnings).

Reply:
> we could also just move its implementation into the preamble lint rule as well if this helps avoid confusion.

I think this is the right call. As of right now, the official WDL spec has no concept of a preamble or preamble comment. Those are entirely our own inventions (that we hope to spread), but IMO that makes this function out of place. The CST and AST should be more-or-less devoid of this team's opinions/conventions/etc, and focus on the "letter of the law". `wdl-lint` is where we can put our "spirit of the law" interpretations.

Reply:
Sounds good, I'll remove it and let the lint rule implement it.